### PR TITLE
Flatpak: add manifest

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
See #54 #42 

There are two missing pieces for it to be ready:
- Install the freedesktop files as part of the cmake buildsystem
- Include a metainfo file that describes the app to a store in-tree and install it as part of the freedesktop files https://www.freedesktop.org/software/appstream/docs/sect-Metadata-Application.html & https://github.com/flathub/flathub/wiki/AppData-Guidelines

The manifest can be built locally with
`flatpak-builder --install repo com.github.blackhole89.notekit.yml --force-clean --user -y` & can be run with `flatpak run com.github.blackhole89.notekit`

### CODE REVIEW

**Comment by sp1ritCS** [comment]
> Install the freedesktop files as part of the cmake buildsystem

I'm currently working a meson buildfile https://github.com/blackhole89/notekit/pull/58 (ready to merge, as https://github.com/NanoMichael/cLaTeXMath/pull/35 is merged), meson installs the freedesktop files and uses cLaTeXMath as shared library (this flatpak package is missing it completely).

**Comment by bilelmoussaoui** [comment]
> > Install the freedesktop files as part of the cmake buildsystem
> 
> I'm currently working a meson buildfile #58 (ready to merge, as [NanoMichael/cLaTeXMath#35](https://github.com/NanoMichael/cLaTeXMath/pull/35) is merged), meson installs the freedesktop files and uses cLaTeXMath as shared library (this flatpak package is missing it completely).

Nice, let me know once the meson changes are in so I can rebase & update this one.

**Comment by bilelmoussaoui** [comment]
I have included clatexmath using the latest commit (to use meson).

**Comment by sp1ritCS** [comment]
You can test the meson version with that:
```diff
diff --git a/com.github.blackhole89.notekit.yaml b/com.github.blackhole89.notekit.yaml
index 44cee80..6c3a4f0 100644
--- a/com.github.blackhole89.notekit.yaml
+++ b/com.github.blackhole89.notekit.yaml
@@ -112,7 +112,8 @@ modules:
         commit: b09ab26129498b7ddf787bbf86894db52b92ffb5
         
   - name: notekit
-    buildsystem: cmake
+    buildsystem: meson
     sources: 
-      - type: dir
-        path: .
+      - type: git
+        url: https://github.com/sp1ritCS/notekit.git
+        branch: meson

```

Some notes: 
- notekit is missing socket=session-bus (`Failed to register: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: org.freedesktop.DBus.Error.ServiceUnknown`)
- on launch: glibmm throws a critical warning (`(process:2): glibmm-CRITICAL **: 22:15:18.511: Class::register_derived_type(): base_query.type_name is NULL.`)
- /usr/share/notekit/data/latex is missing ➡ latex wont render as its fonts are missing. I'll submit a patch to my meson buildfile in clatexmath that installs the contents of the "res/" folder in `/usr/share/clatexmath`. I'm not sure how we should continue from there, either:
  - let notekit use latex fonts from `/usr/share/clatexmath`
  - symlinking `/usr/share/notekit/data/latex` to `/usr/share/clatexmath`
  - or continue doing it the way that we currently use, and copy the contents of `res/` to `data/latex`.  
- `~/.local/share/notekit` (where notes are saved by default) appears to be inside of the sandbox. It might be a good idea to use `/.local/share/notekit` in the users home directory.

**Comment by bilelmoussaoui** [comment]
>  notekit is missing socket=session-bus (Failed to register: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: org.freedesktop.DBus.Error.ServiceUnknown)

No, the app-id is wrongly set. See my commit here. You don't need session-bus at all.

>on launch: glibmm throws a critical warning ((process:2): glibmm-CRITICAL **: 22:15:18.511: Class::register_derived_type(): base_query.type_name is NULL.)

No idea what could cause this, not sure if it's related to the flatpak itself.

> ~/.local/share/notekit (where notes are saved by default) appears to be inside of the sandbox. It might be a good idea to use /.local/share/notekit in the users home directory.

No, Flatpak is all about the sandboxing features.

**Comment by sp1ritCS** [comment]
> No, Flatpak is all about the sandboxing features.

Well yes, but I think a user should have access to his/her Notes. (Especially as they are just a directory hierarchy with markdown files. 

cloud-sync would be very hard if the files are only stored in the sandbox.

**Comment by bilelmoussaoui** [comment]
> > No, Flatpak is all about the sandboxing features.
> 
> Well yes, but I think a user should have access to his/her Notes. (Especially as they are just a directory hierarchy with markdown files.
> 
> cloud-sync would be very hard if the files are only stored in the sandbox.

A simple export button would work just fine as well :)

**Comment by sp1ritCS** [comment]
> A simple export button would work just fine as well :)

Yeah, if I want to send one file to someone else. But again, what about cloud-sync.

**Comment by LyesSaadi** [comment]
Speaking of Flatpak, should we work on an AppData file?

**Comment by sp1ritCS** [comment]
@LyesSaadi AFAIK flat~~pak~~hub now needs metainfo files and not appstream ones. (They are very similar tho)

**Comment by bilelmoussaoui** [comment]
> > A simple export button would work just fine as well :)
> 
> Yeah, if I want to send one file to someone else. But again, what about cloud-sync.

That's an issue up to you upstream to solve. I just contributed the Flatpak manifest.

**Comment by bilelmoussaoui** [comment]
> @LyesSaadi AFAIK flat~pak~hub now needs metainfo files and not appstream ones. (They are very similar tho)

They are the same. The specs suggests using .metainfo.xml for a while instead of .appdata.xml, it's just old tools & old habits.

**Comment by LyesSaadi** [comment]
> Speaking of Flatpak, should we work on an AppData file?

And... nearly done #61 !

**Comment by bilelmoussaoui** [comment]
I went ahead and also added tests to validate the desktop & metainfo files. Flatpak can run the tests inside the sandbox & the SDK has the required executables in already so they should be ran automatically (along with any ninja tests provided through meson). This PR should be ready I guess.

**Comment by sp1ritCS** [comment]
It might be a good idea to move `com.github.blackhole89.notekit.yaml` into a seperate subdirectory (maybe `pkgs/flatpak/`). Or does the flatpak `dir` source not support parent directories?

**Comment by bilelmoussaoui** [comment]
It does. Let's see if the CI passes & will update the PR accordingly

**Comment by sp1ritCS** [comment]
LGTM

**Comment by bilelmoussaoui** [comment]
Once a release is made (don't forget to add a release tag please to the metainfo), the manifest can be used to publish it on Flathub with just a small change to make it use the released tarball instead of the the current directory.

**Comment by LyesSaadi** [comment]
Just tested the Flatpak:

Latex fails:
```
/usr/include/c++/10.2.0/bits/shared_ptr_base.h:1010: std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::element_type& std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::operator*() const [with _Tp = tex::Dummy; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic; bool <anonymous> = false; bool <anonymous> = false; std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::element_type = tex::Dummy]: Assertion '_M_get() != nullptr' failed.
```

And I forgot to remove 2 <em/> block from the metainfo (sorry :sweat:!). Weirdly, appstream-util didn't catch those two!

**Comment by bilelmoussaoui** [comment]
Does it require having `latex` installed? if that's the case, there's a SDK extension that provides the whole latex stuff, but it should definitely be opt-in (as in not installed by default) as it's quiet heavy.

**Comment by LyesSaadi** [comment]
> Does it require having `latex` installed? if that's the case, there's a SDK extension that provides the whole latex stuff, but it should definitely be opt-in (as in not installed by default) as it's quiet heavy.

No, it (hopefully) doesn't, this seems to be a linker/compilation error. Maybe the way it is handled with the symlink. Does it work on Arch Linux with the meson symlinking?

**Comment by bilelmoussaoui** [comment]
> > Does it require having `latex` installed? if that's the case, there's a SDK extension that provides the whole latex stuff, but it should definitely be opt-in (as in not installed by default) as it's quiet heavy.
> 
> No, it (hopefully) doesn't. This seems to be a linker/compilation error. Maybe the way it is handled with the symlink? Does it work on Arch Linux with the meson symlinking?

I have no idea, I would prefer if that issues is solved separately as I don't think there's anything broken with the flatpak itself

**Comment by sp1ritCS** [comment]
> Does it work on Arch Linux with the meson symlinking?

Works on openSUSE, clean podman container.

I couln't test flatpak yet, as I've realized that I cant run graphical flatpak applications from a container.
```
(flatpak run:22153): dconf-CRITICAL **: 22:19:04.300: unable to create file '/home/florian/.cache/dconf/user': Permission denied.  dconf will not work properly.

(flatpak run:22153): dconf-CRITICAL **: 22:19:04.300: unable to create file '/home/florian/.cache/dconf/user': Permission denied.  dconf will not work properly.
bwrap: Can't find source path /tmp/.X11-unix/X0: No such file or directory
```

**Comment by bilelmoussaoui** [comment]
Why run a container inside a container? :)

**Comment by sp1ritCS** [comment]
> Just tested the Flatpak:
> 
> Latex fails:
> 
> ```
> /usr/include/c++/10.2.0/bits/shared_ptr_base.h:1010: std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::element_type& std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::operator*() const [with _Tp = tex::Dummy; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic; bool <anonymous> = false; bool <anonymous> = false; std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::element_type = tex::Dummy]: Assertion '_M_get() != nullptr' failed.
> ```

Fails here too

> I have no idea, I would prefer if that issues is solved separately as I don't think there's anything broken with the flatpak itself

Well, it works in a clean container, it doesn't work as flatpak. Maybe there is an issue with the versions of the dependencies? Other than that I have no idea.

**Comment by bilelmoussaoui** [comment]
Version of which dependencies exactly? as of today, the buildsystem you're using doesn't enforce any specific versions and I don't speak C++ so I have no idea what the error is about.

**Comment by blackhole89** [comment]
> Why run a container inside a container? :)

Well, you might not trust the inner containerisation system :)

Of course, you might actually not trust any containerisation systems at all, in which case your only option is to put another container around that, and so on ad infinitum.

> Latex fails

That error seems to be saying that it is trying to dereference an STL smart pointer, but the pointer it is wrapping is NULL. Would it be possible to see some more of the call stack?

(Does the error occur on startup or only when trying to render some LaTeX?)

> Does it require having latex installed?

I believe it shouldn't (everything is in the res/ folder that should have been moved to /usr/share/clatexmath), but not 100% confidence here as I've never tested it on a system without. Maybe the error indicates that for some reason or another, it failed to read those resources from wherever they did wind up getting put. (Was the symlink created properly? Does it point at the right thing (as opposed to, say, a folder _containing_ the thing it's supposed to point at, or conversely a subfolder of the thing it's supposed to point at)? Is whatever API cLaTeXMath uses for loading files in fact capable of following symlinks?)

**Comment by sp1ritCS** [comment]
yes, it is capable of following symlinks - as it works on a container

---

the symlink links to /app/share/clatexmath, dunno if this is the correct location

File: pkgs/flatpak/com.github.blackhole89.notekit.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+          - type: archive
+            url: https://download.gnome.org/sources/gtksourceview/3.24/gtksourceview-3.24.8.tar.xz
+            sha256: 1e9bb8ff190db705deb916dd23ff681f0e8803aec407bf0fd64c7e615ac436fe
+
+  - name: cLaTeXMath
+    buildsystem: meson
+    sources:
+      - type: git
+        url: https://github.com/NanoMichael/cLaTeXMath
+        commit: b09ab26129498b7ddf787bbf86894db52b92ffb5
```
Comment (sp1ritCS) [inline pkgs/flatpak/com.github.blackhole89.notekit.yaml:112]: the commit sha hasn't been updated to include the data,
`44ea33777c9f3ffde8f1d2149ce30362accebe63`

**Comment by sp1ritCS** [comment]
I don't know, now its even more obscure - as soon as the rendered latex supersedes the size of one char it continues to crash with the same error.

example:
$a$, or even $\frac{1}{2}$ works;
$aa$ doesnt

File: pkgs/flatpak/com.github.blackhole89.notekit.yaml (hunk 1 above)
Comment (bilelmoussaoui) [reply pkgs/flatpak/com.github.blackhole89.notekit.yaml:112]: Feel free to finish the remaining parts and merge it.

**Comment by LyesSaadi** [comment]
> I don't know, now its even more obscure - as soon as the rendered latex supersedes the size of one char it continues to crash with the same error.
> 
> example:
> $a$, or even $\frac{1}{2}$ works;
> $aa$ doesnt

I just tested it, what you sent happens for Fedora/Arch/OpenSUSE. But on Flatpak it fails whatever is the string, surely because the symlinking is broken as you pointed out!

The error, though, is exactly the same.

**Comment by LyesSaadi** [comment]
> That error seems to be saying that it is trying to dereference an STL smart pointer, but the pointer it is wrapping is NULL. Would it be possible to see some more of the call stack?
>
> (Does the error occur on startup or only when trying to render some LaTeX?)

I'm finally able to provide the backtrace, since I've just got through the creation of a new cLaTeXMath-compatible RPM thanks to @sp1ritCS !

It does not happen at startup (though I guess it will, if we save the problematic LaTeX), it only happens when it wants to render it.

```
(gdb) backtrace
#0  0x00007ffff6539365 in raise () at /lib64/libc.so.6
#1  0x00007ffff65228a4 in abort () at /lib64/libc.so.6
#2  0x00007ffff6aaabb8 in std::__replacement_assert(char const*, int, char const*, char const*)
    (__file=__file@entry=0x7ffff6b59758 "/usr/include/c++/10/bits/shared_ptr_base.h", __line=__line@entry=1010, __function=__function@entry=0x7ffff6b59788 "std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::element_type& std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::operator*() const [with _Tp = tex::Dummy; __gnu_cxx::_Lock_p"..., __condition=__condition@entry=0x7ffff6b59b78 "_M_get() != nullptr") at /usr/include/c++/10/x86_64-redhat-linux/bits/c++config.h:2560
#3  0x00007ffff6ab0cc8 in std::__shared_ptr_access<tex::Dummy, (__gnu_cxx::_Lock_policy)2, false, false>::operator*() const (this=<optimized out>)
    at /usr/include/c++/10/bits/shared_ptr_base.h:1010
#4  std::__shared_ptr_access<tex::Dummy, (__gnu_cxx::_Lock_policy)2, false, false>::operator*() const (this=0x555555cfa730)
    at /usr/include/c++/10/bits/shared_ptr_base.h:1008
#5  tex::RowAtom::createBox(tex::TeXEnvironment&) (this=0x555555cfa6f0, env=...)
    at ../src/atom/atom_basic.cpp:483
#6  0x00007ffff6b2c579 in tex::TeXRenderBuilder::build(std::shared_ptr<tex::Atom> const&) (this=0x55555588f9d0, fc=<optimized out>) at ../src/render.cpp:135
#7  0x000055555559037d in CLatexWidget::CLatexWidget(Glib::RefPtr<Gdk::Window>, Glib::ustring, Gdk::RGBA)
    (fg=..., text=..., wnd=..., this=0x555555c93700, this=<optimized out>, wnd=..., --Type <RET> for more, q to quit, c to continue without paging--
text=..., fg=...) at /usr/include/c++/10/bits/allocator.h:162
#8  CNotebook::RenderToWidget(Glib::ustring, Gtk::TextIter&, Gtk::TextIter&)
    (end=..., start=..., wtype=..., this=0x55555576b6d0) at ../notebook.cpp:516
#9  CNotebook::on_leave_region(Gtk::TextIter&, Gtk::TextIter&)
    (this=0x55555576b6d0, start=..., end=...) at ../notebook.cpp:650
#10 0x000055555558bb5d in CNotebook::on_highlight_updated(Gtk::TextIter&, Gtk::TextIter&) (this=0x55555576b6d0, start=..., end=...) at ../notebook.cpp:391
#11 0x00007ffff6c911d2 in (anonymous namespace)::Buffer_signal_highlight_updated_callback(_GtkSourceBuffer*, _GtkTextIter*, _GtkTextIter*, void*) ()
    at /lib64/libgtksourceviewmm-3.0.so.0
#12 0x00007ffff5b89b10 in ffi_call_unix64 () at /lib64/libffi.so.6
#13 0x00007ffff5b890a3 in ffi_call () at /lib64/libffi.so.6
#14 0x00007ffff6e0d8cd in g_cclosure_marshal_generic ()
    at /lib64/libgobject-2.0.so.0
#15 0x00007ffff6e07e2a in g_closure_invoke () at /lib64/libgobject-2.0.so.0
#16 0x00007ffff6e314f2 in signal_emit_unlocked_R.isra.0 ()
    at /lib64/libgobject-2.0.so.0
#17 0x00007ffff6e25012 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
#18 0x00007ffff6e25449 in g_signal_emit_by_name () at /lib64/libgobject-2.0.so.0
#19 0x00007ffff5ac7e76 in refresh_range () at /lib64/libgtksourceview-3.0.so.1
#20 0x00007ffff5ace473 in update_syntax () at /lib64/libgtksourceview-3.0.so.1
#21 0x00007ffff5ad0001 in first_update_callback ()
    at /lib64/libgtksourceview-3.0.so.1
--Type <RET> for more, q to quit, c to continue without paging--
#22 0x00007ffff6fcb0c9 in gdk_threads_dispatch () at /lib64/libgdk-3.so.0
#23 0x00007ffff6d1432b in g_idle_dispatch () at /lib64/libglib-2.0.so.0
#24 0x00007ffff6d1526f in g_main_context_dispatch () at /lib64/libglib-2.0.so.0
#25 0x00007ffff6d67a68 in g_main_context_iterate.constprop ()
    at /lib64/libglib-2.0.so.0
#26 0x00007ffff6d12643 in g_main_context_iteration () at /lib64/libglib-2.0.so.0
#27 0x00007ffff637d5bd in g_application_run () at /lib64/libgio-2.0.so.0
#28 0x0000555555577e27 in main(int, char**)
    (argc=<optimized out>, argv=<optimized out>)
    at /usr/include/glibmm-2.4/glibmm/refptr.h:259
```

(Btw, @sp1ritCS, here are my (Fedora Guidelines-compliant) [spec files](https://gitlab.com/LyesSaadi/spec/-/tree/master/notekit) if you ever want to check them :3)

**Comment by sp1ritCS** [comment]
> I just tested it, what you sent happens for Fedora/Arch/OpenSUSE. But on Flatpak it fails whatever is the string, surely because the symlinking is broken as you pointed out!

Wut? That what I sent occurred for me on Flatpak. On openSUSE it works flawlessly
grafik

I've also tried to replace `~/.local/share/flatpak/app/com.github.blackhole89.notekit/current/active/files/share/notekit/data/latex` with the `res/` directory from cLaTeXMath directly w/o symlinking it. Same error (Unless that it not the correct location for swapping out files inside the flatpak namespace)

I'm hoping that @NanoMichael may already encountered this issue and knows how to fix it.

> for Fedora/Arch/OpenSUSE
did you just run meson / ninja / install on every distro or did you use the packaged versions? Because I haven't upgraded my AUR package to meson yet ➡ an issue with cLaTeXMath source.

**Comment by LyesSaadi** [comment]
Oh, No! Sorry, I misunderstood your comment based on some wrong assumptions!

Basically, on Fedora I get the same error as on Flatpak. I saw your comment saying that you weren't able to use Flatpak and thought that the errors you got where on OpenSUSE or Arch...

I guess the fact that it fails with 1 or more char is just related to the usage of the old commit sha as you said.

**Comment by sp1ritCS** [comment]
> Basically, on Fedora I get the same error as on Flatpak. I saw your comment saying that you weren't able to use Flatpak and thought that the errors you got where on OpenSUSE or Arch...

that should rule out a wrong version of a dependency (as fedora 33 was released between leap 15.2 and tumbleweed) 

anyways, that should make it a lot easier to test, what the issue might be. can you try to relink the symlink? or replace the symlink with an seclink or a copy of the folder?

**Comment by LyesSaadi** [comment]
Um... Replacing the symlink with a copy of the folder crashes...

So, this isn't the symlink...

**Comment by sp1ritCS** [comment]
> So, this isn't the symlink...

Gussed as much :/

Another thing that might be worth trying is to build the cLaTeXMath demo using meson. If i'm not mistaken, it should be enough to replace `library` with `executable` in its meson.build file.

**Comment by blackhole89** [comment]
Hm. The relevant parts of the cLaTeXMath code are (with the crashing line being the last one):

```cpp
    sptr<Dummy> atom(new Dummy(at));
    // (...)
    sptr<Atom> nextAtom(nullptr);
    if (i < e) nextAtom = _elements[i + 1];
    change2Ord(&(*atom), &(*_previousAtom), &(*nextAtom));
```

My first thought was that this could possibly result in a straight-up null pointer deref if the `i<e` condition fails, but actually, looking at frame #4 and up, it seems like we're looking at a shared_ptr to tex::Dummy, so only the first argument of change2Ord is implicated. That, however, seems very strange, as `atom` was just freshly allocated a few lines prior. At that point, the only options are an out-of-memory error (operator new fails) or stack corruption. Does Flatpak impose RAM usage limits? Could you go up to frame #5 and look at what `atom` (resp. `*atom`) evaluates to? Maybe also worth breaking at atom_basic.cpp:478 and stepping through while looking at what's going on.

(Alternatively, is there any output to stdout happening before the crash? (Maybe cLaTeXMath complains about not being able to find some file.))

(Alternatively², do you think you could upload the crashing image you obtained? I might be able to debug it myself too.)

**Comment by sp1ritCS** [comment]
> Basically, on Fedora I get the same error as on Flatpak. I saw your comment saying that you weren't able to use Flatpak and thought that the errors you got where on OpenSUSE or Arch...

Works flawlessly under Fedora as well (btw, why does meson have vim-filesystem as dependency?)
image

I'd love to try your package, but I don't know how do build it :/.

---

> I might be able to debug it myself too.

Idk, you could try if you give flatpak the allow=devel permission. Then you could debug inside the flatpak namespace (flatpak enter (the pid you got from flatpak ps) bash). Otherwise I think gnome-builder supports debugging using flatpak manifests

**Comment by sp1ritCS** [comment]
Update: This is an issue with cLaTeXMath - it has "nothing" to do with notekit. Maybe we should move it to an issue request on their repo or on the flatpak repo/discourse maybe.

The cLaTeXMath Demo has the same issue on Flatpak. Apply this patch on https://github.com/NanoMichael/cLaTeXMath, and build it: `flatpak-builder --force-clean _build --user --install com.github.nanomichael.clatexmath.yaml`


https://gist.github.com/sp1ritCS/b71b2656cfa35e362ff2524e8015e74c

It'll fail if the rendered string is larger than one.
Entire log:
```
Gtk-Message: 18:00:09.402: Failed to load module "canberra-gtk-module"
Gtk-Message: 18:00:09.447: Failed to load module "canberra-gtk-module"
Gtk-Message: 18:00:09.449: Failed to load module "canberra-gtk-module"

(clatexmath:2): glibmm-CRITICAL **: 18:00:09.469: Class::register_derived_type(): base_query.type_name is NULL.
/usr/include/c++/10.2.0/bits/shared_ptr_base.h:1010: std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::element_type& std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::operator*() const [with _Tp = tex::Dummy; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic; bool <anonymous> = false; bool <anonymous> = false; std::__shared_ptr_access<_Tp, _Lp, <anonymous>, <anonymous> >::element_type = tex::Dummy]: Assertion '_M_get() != nullptr' failed.
```

**Comment by LyesSaadi** [comment]
> I'd love to try your package, but I don't know how to build it :/.

I've done it for you:
```
sudo dnf install https://lyessaadi.fedorapeople.org/notekit/notekit-0-2.fc34.x86_64.rpm https://lyessaadi.fedorapeople.org/notekit/cLaTeXMath-0-1.fc34.x86_64.rpm
```

---

Maybe is this a permission issue?

**Comment by sp1ritCS** [comment]
```
Error: 
 Problem: conflicting requests
  - nothing provides libc.so.6(GLIBC_2.33)(64bit) needed by notekit-0-2.fc34.x86_64
```

I can install clatexmath library (can you "compile" the header package aswell?), it cannot install notekit (clean fedora 33 install)

**Comment by blackhole89** [comment]
> (clatexmath:2): glibmm-CRITICAL **: 18:00:09.469: Class::register_derived_type(): base_query.type_name is NULL.

That's slightly worrying, as I don't remember seeing it in normal operation, but I'm not sure it's related to the problem we're seeing.

> change2Ord(&(*atom), &(*_previousAtom), &(*nextAtom));

Correction: _previousAtom is also an sptr<Dummy>, and in fact if I read the code correctly, it *should* be a pointer to NULL at first.

I have another suspicion, which is that it might be to do with some g++ 10 related change. To my (untrained in C++ rules lawyering) eye, this whole stretch of code actually looks suspicious: if `p` is a shared_ptr<T> to NULL, clearly the code is written with the assumption that `&(*p)` can be evaluated to a T* that equals NULL, but wouldn't compositionality suggest that on some level, we already have illicitly dereferenced a NULL pointer when computing `*p`, and this is not redeemed by the circumstance that we "only" take its address, rather than its value, afterwards? Indeed, my local implementation of `__shared_ptr_access` reads,
```cpp
      element_type&
      operator*() const noexcept
      {
	__glibcxx_assert(_M_get() != nullptr);
	return *_M_get();
      }
```
and I would expect that assertion failing being exactly the crash that we are seeing. Considering that, it's quite puzzling to me that on my system (g++ 9.3.0), the following compiles and runs just fine, with or without `-g`:
```cpp
#include <stdio.h>
#include <memory>

struct Lol {
    int lol;
};

int main(int argc, char* argv[])
{
    std::shared_ptr<Lol> p1(new Lol());
    std::shared_ptr<Lol> p2(nullptr);

    printf("%p %p\n",&(*p1),&(*p2));
    return 0;
}
```
If one of you has the g++ version that was used to build the failing clatexmath (10.2.0), could you copypaste this blurb into a new file and try compiling and running it (with and without `-g`, and maybe with some different `-O` levels too)? If any one variant of it fails for you, I think that would imply an answer to what the issue is (as well as suggest a trivial solution, which might at most be cumbersome if @NanoMichael used that "convert smart pointer to dumb" trick in many places).

**Comment by sp1ritCS** [comment]
> > (clatexmath:2): glibmm-CRITICAL **: 18:00:09.469: Class::register_derived_type(): base_query.type_name is NULL.
>
> That's slightly worrying, as I don't remember seeing it in normal operation, but I'm not sure it's related to the problem we're seeing.

https://github.com/blackhole89/notekit/pull/60#issuecomment-728340920
> on launch: glibmm throws a critical warning ((process:2): glibmm-CRITICAL **: 22:15:18.511: Class::register_derived_type(): base_query.type_name is NULL.)

---

> If one of you has the g++ version that was used to build the failing clatexmath (10.2.0), could you copypaste this blurb into a new file and try compiling and running it 

I have a bit new gcc (g++ (SUSE Linux) 10.2.1 20201028 [revision a78cd759754c92cecbf235ac9b447dcdff6c6e2f])
nothing failed tho

**Comment by LyesSaadi** [comment]
> If one of you has the g++ version that was used to build the failing clatexmath (10.2.0), could you copypaste this blurb into a new file and try compiling and running it (with and without -g, and maybe with some different -O levels too)?

Doing that, I consistently have an output with a random hex address and "nil":
```
0xr1nd0m (nil)
```
Which is what is expected if I understand your code correctly, right?

Do you maybe want me to send the build warnings with GCC 10?

---

@sp1ritCS, sorry, I sent RPMs compiled on Fedora 34 (rawhide), and not Fedora 33. I just compiled them for you to test right now. To install the requested packages, do this:
```
sudo dnf remove cLaTeXMath
sudo dnf install https://lyessaadi.fedorapeople.org/notekit/cLaTeXMath-0-1.fc33.x86_64.rpm https://lyessaadi.fedorapeople.org/notekit/cLaTeXMath-devel-0-1.fc33.x86_64.rpm https://lyessaadi.fedorapeople.org/notekit/notekit-0-2.fc33.x86_64.rpm
```

Also, I've also uploaded [more packages](https://lyessaadi.fedorapeople.org/notekit/) if you want to install them (the debuginfo for gdb & the src.rpm).

**Comment by blackhole89** [comment]
@LyesSaadi Yeah, that's what you should be seeing. Try it with -D_GLIBCXX_DEBUG though, and you'll see the same crash.

I'm currently preparing a pull request against cLaTeXMath to fix the problem (or at least this instance of it; not sure if the same thing is happening anywhere else in the code).

**Comment by LyesSaadi** [comment]
> @LyesSaadi Yeah, that's what you should be seeing. Try it with -D_GLIBCXX_DEBUG though, and you'll see the same crash.

Yes! Indeed! This seems to be the issue!

Fedora builds with `-D_GLIBCXX_ASSERTIONS` which also causes it to fail :tada:!

**Comment by sp1ritCS** [comment]
> Fedora builds with -D_GLIBCXX_ASSERTIONS which also causes it to fail tada!

same on opensuse - so it's probably just a thing that the %meson macro does different between the distros (and flatpak-builder)

**Comment by LyesSaadi** [comment]
Manually removing `-D_GLIBCXX_ASSERTIONS` works :tada:! This was it!

image

**Comment by blackhole89** [comment]
For what it's worth, I made a PR over at cLaTeXMath for this, but maybe leaving _GLIBCXX_ASSERTIONS off here is also not a bad idea independently (...isn't having the assertions up kind of slow?).

Is that the last thing that needs to be done before this PR can be merged?

**Comment by sp1ritCS** [comment]
Jup, Flatpak is working now 
image

Test using:
```diff
diff --git a/pkgs/flatpak/com.github.blackhole89.notekit.yaml b/pkgs/flatpak/com.github.blackhole89.notekit.yaml
index 0fb27fb..8039d72 100644
--- a/pkgs/flatpak/com.github.blackhole89.notekit.yaml
+++ b/pkgs/flatpak/com.github.blackhole89.notekit.yaml
@@ -108,8 +108,9 @@ modules:
     buildsystem: meson
     sources:
       - type: git
-        url: https://github.com/NanoMichael/cLaTeXMath
-        commit: 44ea33777c9f3ffde8f1d2149ce30362accebe63
+        url: https://github.com/blackhole89/cLaTeXMath
+        branch: patch-2
+        commit: 74d3226aaf1d35bdd51c43f3e8dfcba22548f73b
         
   - name: notekit
     buildsystem: meson
```

**Comment by blackhole89** [comment]
The PR over at cLaTeXMath got merged now, so I assume we're just waiting for the commit ID for it to be updated.

**Comment by bilelmoussaoui** [comment]
> The PR over at cLaTeXMath got merged now, so I assume we're just waiting for the commit ID for it to be updated.

Done :)

**Comment by bilelmoussaoui** [comment]
Should be ready to go I guess :D

**Comment by sp1ritCS** [comment]
>Should be ready to go I guess :D

installed the action build - works. 
grafik

Ready to merge

**Comment by blackhole89** [comment]
Many thanks for everyone's help with this!

I updated the screenshot (and added an additional one to show off the syntax highlighting; this is something that probably needs to be tested separately, though, since it's a fairly brittle feature and I have no idea if it'll work in a container) in the README the other day. Let me know if I also should write a new expository blurb that is better suited for Flatpak stores, or there's anything else of that kind that needs to be done.

**Comment by sp1ritCS** [comment]
> or there's anything else of that kind that needs to be done.

someone needs to submit (and maintain) the application to the flathub repository - or do you think we should wait until we have a better description (and Adwaita screenshots with transparent background)

**Comment by LyesSaadi** [comment]
> or there's anything else of that kind that needs to be done.

Yes! We need to get the first release done :smiley:! I know it might be hard to make the first big step toward a stable release, but please tag a version (after you've written the release notes in the metainfo file, please :stuck_out_tongue:, and maybe improved the marketing, like better description and screenshots) and release it (on GitHub) :tada:! This will allow anyone who would maintain the Flatpak to submit it to Flathub, and for us Package Maintainer to finally have a stable version to stick to :smiley:! For me, it would allow me to submit it to Fedora's official repositories for example.

What will be the first version? `0.1.0` or `1.0.0`! According to how advanced you feel your project is!

I think the next step would be as @sp1ritCS mentioned, to work on a better description and screenshots (if not Adwaita, at least the transparent background through GNOME's screenshot tool)!

**Comment by sp1ritCS** [comment]
> For me, it would allow me to submit it to Fedora's official repositories for example.

What do you do about clatexmath? It also doesn't have a "proper"  versioning. While I could submit a package using commitTimeStamp.commitId/Hash as version to Factory I try to avoid it and I usually refrain to pushing alpha/beta software to Factory.

But the flatpak packages I maintain also have just relasedate as version. So, in theory flatpak/hub does not need to have a "proper" version.

> I think the next step would be as @sp1ritCS mentioned, to work on a better description and screenshots (if not Adwaita, at least the transparent background through GNOME's screenshot tool)!

Or just send the demo file, and someone else (i.e LyesSaadi or me) could take a better screenshot, because I'm unsure if gnome screenshot has those fancy features in debian stable already.

**Comment by blackhole89** [comment]
I don't have the GNOME screenshot tool (using MATE here), and producing transparent-background screenshots indeed may be somewhat cumbersome. Any release at the moment should definitively be 0.1.0 rather than 1.0.0; you can't even ctrl-f text in files yet.

Here are the markdown files I used for the topmost screenshot and syntax highlighting (sorry for the file format, it was driven by Github's whitelist): [demonotes.zip](https://github.com/blackhole89/notekit/files/5668141/demonotes.zip)

**Comment by LyesSaadi** [comment]
Hey ! So I recreated the whole treeview to be exactly like the photo, and here are my photos:

With Adwaita:
image
image

With Adwaita Dark:
image
image


Choose which color variant you prefer, or even include both!

**Comment by sp1ritCS** [comment]
> Choose which color variant you prefer, or even include both!

I think using the first Adwaita one for the readme and have a link for more images below that is the best solution. The metainfo should have pictures every picture.

**Comment by thismarty** [comment]
Hi, devs!  Looking forward to the first release!  Anything a non-C++ coder can do to help?

**Comment by blackhole89** [comment]
Thanks! Well, there's the issue in #72/#77, which is mostly about [Glib regular expressions](https://developer.gnome.org/glib/stable/glib-regex-syntax.html) and (to a lesser extent?) the (plain-C) GtkSourceView syntax highlighting engine, but it's a total slog...

**Pull request merged**